tilepack/disk_outputter.go

```diff
@@ -0,0 +1,108 @@
+package tilepack
+
+import (
+	"errors"
+	"fmt"
+	"github.com/aaronland/go-string/dsn"
+	_ "log"
+	"os"
+	"path/filepath"
+)
+
+type diskOutputter struct {
+	TileOutputter
+	root     string
+	format   string
+	hasTiles bool
+}
+
+func NewDiskOutputter(dsn_str string) (*diskOutputter, error) {
+
+	dsn_map, err := dsn.StringToDSNWithKeys(dsn_str, "root", "format")
+
+	if err != nil {
+		return nil, err
+	}
+	
+	abs_root, err := filepath.Abs(dsn_map["root"])
+	
+	if err != nil {
+		return nil, err
+	}
+
+	o := diskOutputter{
+		root:   abs_root,
+		format: dsn_map["format"],
+	}
+
+	return &o, nil
+}
+
+func (o *diskOutputter) Close() error {
+	return nil
+}
+
+func (o *diskOutputter) CreateTiles() error {
+	if o.hasTiles {
+		return nil
+	}
+
+	info, err := os.Stat(o.root)
+
+	if err != nil {
+
+		if os.IsNotExist(err) {
+
+			err := os.MkdirAll(o.root, 0755)
+
+			if err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+
+	} else {
+
+		if !info.IsDir() {
+			return errors.New("Root is already a file")
+		}
+	}
+
+	o.hasTiles = true
+	return nil
+}
+
+func (o *diskOutputter) Save(tile *Tile, data []byte) error {
+
+	rel_path := fmt.Sprintf("%d/%d/%d.%s", tile.Z, tile.X, tile.Y, o.format)
+	abs_path := filepath.Join(o.root, rel_path)
+
+	root := filepath.Dir(abs_path)
+
+	_, err := os.Stat(root)
+
+	if os.IsNotExist(err) {
+		err = os.MkdirAll(root, 0755)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	fh, err := os.OpenFile(abs_path, os.O_RDWR|os.O_CREATE, 0644)
+
+	if err != nil {
+		return err
+	}
+
+	defer fh.Close()
+
+	_, err = fh.Write(data)
+
+	if err != nil {
+		return err
+	}
+
+	return fh.Close()
```

Comment:
This call to `fh.Close()` twice is super confusing to me. I know it works; we get the error from the call in the final `return`, and the safety of closing if we exit earlier due to some other error. However, it is an error to close a file twice - we'll just never see that error because `defer` throws it away.

How would you feel about something like this instead?

```golang
	_, err = fh.Write(data)

	if err != nil {
		fh.Close()
		return err
	}

	return fh.Close()
```

Reply:
Thanks! I didn't think it was a big problem, but it caught my eye and confused me.

When I tested it out, [I got an error](https://play.golang.org/p/bVu-MR-3iPr). When the second close is in a `defer`, it effectively acts like `_ = f.Close()` and throws the error away. Which kind of makes it a no-op, unless `f.Close()` grows some noticeable side-effects in the future.

Reply:
Mercifully, the Go people are pretty good about a) not doing that b) giving people plenty of time and warning if that kind of change will happen

Reply:
Indeed! I wasn't so much worried about the Go people, as by myself at a later date replacing that with something implementing `io.WriteCloser` which did have side-effects.